omp/solver/cb_gmres_kernels.cpp

```diff
@@ -0,0 +1,511 @@
+                         size_type iter, const stopping_status *stop_status)
+{
+    const remove_complex<ValueType> eta = 1.0 / sqrt(2.0);
+#pragma omp declare reduction(add:ValueType : omp_out = omp_out + omp_in)
+#pragma omp declare reduction(addnc : remove_complex<ValueType> : omp_out = omp_out + omp_in)
+#pragma omp declare reduction(infnc : remove_complex<ValueType> : omp_out = (omp_out >= omp_in? omp_out: omp_in))
+
+    for (size_type i = 0; i < next_krylov_basis->get_size()[1]; ++i) {
+        if (stop_status[i].has_stopped()) {
+            continue;
+        }
+
+        remove_complex<ValueType> nrm = zero<remove_complex<ValueType>>();
+#pragma omp parallel for reduction(addnc : nrm)
+        for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+            nrm += squared_norm(next_krylov_basis->at(j, i));
+        }
+        arnoldi_norm->at(0, i) = nrm * eta;
+        // nrmP = norm(next_krylov_basis)
+#ifdef TIMING
+        auto start_1 = std::chrono::steady_clock::now();
+#endif
+#pragma omp parallel for
+        for (size_type k = 0; k < iter + 1; ++k) {
+            ValueType hessenberg_iter_entry = zero<ValueType>();
+            for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+                hessenberg_iter_entry +=
+                    next_krylov_basis->at(j, i) * krylov_bases(k, j, i);
+            }
+            hessenberg_iter->at(k, i) = hessenberg_iter_entry;
+        }
+#ifdef TIMING
+        auto time_1 = std::chrono::steady_clock::now() - start_1;
+        std::cout << "time_1(" << iter << ") = "
+                  << std::chrono::duration_cast<double_seconds>(time_1).count()
+                  << std::endl;
+#endif
+        // for i in 1:iter
+        //     hessenberg(iter, i) = next_krylov_basis' * krylov_bases(:, i)
+        // end
+#ifdef TIMING
+        auto start_2 = std::chrono::steady_clock::now();
+#endif
+        for (size_type k = 0; k < iter + 1; ++k) {
+#pragma omp parallel for
+            for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+                next_krylov_basis->at(j, i) -=
+                    hessenberg_iter->at(k, i) * krylov_bases(k, j, i);
+            }
+        }
+#ifdef TIMING
+        auto time_2 = std::chrono::steady_clock::now() - start_2;
+        std::cout << "time_2(" << iter << ") = "
+                  << std::chrono::duration_cast<double_seconds>(time_2).count()
+                  << std::endl;
+        std::cout
+            << "time_1 / time_2(" << iter << ") = "
+            << std::chrono::duration_cast<double_seconds>(time_1).count() /
+                   std::chrono::duration_cast<double_seconds>(time_2).count()
+            << std::endl;
+#endif
+        // for i in 1:iter
+        //     next_krylov_basis  -= hessenberg(iter, i) * krylov_bases(:, i)
+        // end
+        // ValueType nrm = zero<ValueType>();
+        nrm = zero<remove_complex<ValueType>>();
```

Comment:
Again, storing the `rc_vtype` would probably help in terms of coding style since it's used everywhere.